File: src/cpp/core/Vertex.cpp
#include "Vertex.h"
#include "HalfEdge.h"
#include "Face.h"

std::vector<Vertex*> Vertex::star() const {
    std::vector<Vertex*> neighbors;
    if (!halfedge) return neighbors;
    
    HalfEdge* he = halfedge;
    do {
        neighbors.push_back(he->next->vertex);
        he = he->twin->next;
    } while (he != halfedge);
    
    return neighbors;
}

int Vertex::degree() const {
    if (!halfedge) return 0;
    
    int count = 0;
    HalfEdge* he = halfedge;
    do {
        count++;
        he = he->twin->next;
    } while (he != halfedge);
    
    return count;
}

bool Vertex::isBoundary() const {
    if (!halfedge) return true;
    
    HalfEdge* he = halfedge;
    do {
        if (!he->face) return true;
        he = he->twin->next;
    } while (he != halfedge);
    
    return false;
}

std::vector<HalfEdge*> Vertex::outgoingHalfEdges() const {
    std::vector<HalfEdge*> halfedges;
    if (!halfedge) return halfedges;
    
    HalfEdge* he = halfedge;
    do {
        halfedges.push_back(he);
        he = he->twin->next;
    } while (he != halfedge);
    
    return halfedges;
}

std::vector<Face*> Vertex::adjacentFaces() const {
    std::vector<Face*> faces;
    if (!halfedge) return faces;
    
    HalfEdge* he = halfedge;
    do {
        if (he->face) {
            faces.push_back(he->face);
        }
        he = he->twin->next;
    } while (he != halfedge);
    
    return faces;
}

File: src/csharp/Core/Vertex.cs
using System;
using System.Collections.Generic;
using System.Numerics;

namespace DDGCompanion.Core
{
    public class Vertex
    {
        public Vector3 Position { get; set; }
        public HalfEdge? HalfEdge { get; set; }
        public int Index { get; set; } = -1;
        
        public Vertex(Vector3 position)
        {
            Position = position;
        }
        
        public List<Vertex> Star()
        {
            var neighbors = new List<Vertex>();
            if (HalfEdge == null) return neighbors;
            
            var he = HalfEdge;
            var visited = new HashSet<int>();
            while (he != null && visited.Add(he.Index))
            {
                if (he.Vertex != null)
                {
                    neighbors.Add(he.Vertex);
                }

                if (he.Twin?.Next == null) break;
                he = he.Twin.Next;
                if (he == HalfEdge) break;
            }
            
            return neighbors;
        }
        
        public int Degree()
        {
            if (HalfEdge == null) return 0;
            int count = 0;
            var he = HalfEdge;
            var visited = new HashSet<int>();
            while (he != null && visited.Add(he.Index))
            {
                count++;

                if (he.Twin?.Next == null) break;
                he = he.Twin.Next;
                if (he == HalfEdge) break;
            }
            return count;
        }
        
        public bool IsBoundary()
        {
            if (HalfEdge == null) return true;
            var he = HalfEdge;
            var visited = new HashSet<int>();
            while (he != null && visited.Add(he.Index))
            {
                if (he.Face == null || he.Twin == null || he.Twin.Face == null || he.Twin.Next == null) return true;
                he = he.Twin.Next;
                if (he == HalfEdge) break;
            }
            return false;
        }
    }
}
